clients/java/signalr/src/main/java/com/microsoft/signalr/HubConnection.java

```diff
@@ -187,15 +189,22 @@ void setTickRate(Duration tickRate) {
 
     private void timeoutHandshakeResponse(long timeout, TimeUnit unit) {
         ScheduledExecutorService scheduledThreadPool = Executors.newSingleThreadScheduledExecutor();
-        scheduledThreadPool.schedule(() -> handshakeResponseFuture.completeExceptionally(
-                new TimeoutException("Timed out waiting for the server to respond to the handshake message.")), timeout, unit);
+        scheduledThreadPool.schedule(() -> {
+            // If onError is called on a completed subject the global error handler is called
+            // REVIEW: Should we lock?
+            if (!(handshakeResponseSubject.hasComplete() || handshakeResponseSubject.hasThrowable()))
+            {
+                handshakeResponseSubject.onError(
+                    new TimeoutException("Timed out waiting for the server to respond to the handshake message."));
+            }
+        }, timeout, unit);
     }
 
-    private CompletableFuture<NegotiateResponse> handleNegotiate(String url) {
+    private Single<NegotiateResponse> handleNegotiate(String url) {
         HttpRequest request = new HttpRequest();
         request.addHeaders(this.headers);
 
-        return httpClient.post(Negotiate.resolveNegotiateUrl(url), request).thenCompose((response) -> {
+        return httpClient.post(Negotiate.resolveNegotiateUrl(url), request).flatMap((response) -> {
```

Comment:
If you're returning `Single.just` you don't need `flatMap`, you can just use `map` and return `negotiateResponse`. `flatMap` is used when you need to perform another **asynchronous** operation when the previous operation completes.